Remove kube-proxy; Fully migrate to client-go v2.0.0

euank opened the pull request:
This isn't fully tested yet and shouldn't be merged quite yet, but I don't know of any specific issues with it.

The TPR certificate watch thingy does seem to work fine, and that was one of the more fiddly bits.

This was a fairly rote translation from the kubectl-proxy requests to proper client requests. I also think that there's room for code-layout / splitting-into-packages / etc improvements, but I'd rather keep this change smaller by avoiding significant refactoring.

Fixes #14.

luna-duclos commented:
I'd be in favor of committing the whole vendor tree, even though it's big, I have too many bad experiences with part of the locked tree changing or becoming unavailable due to force-pushes

luna-duclos commented:
I'd also like a few people to test this before we merge it, could some people please try this and thumbs up if it works, and reply if it doesn't ?

euank commented:
I updated the code to be a bit cleaner around watching.
This also hopefully fixed a bug I hit where some specific watch error would result in tremendous amounts of logspam.

I'll happily commit the vendor tree, but let's either do that after review/before merge, or in a different PR. Vendoring k8s sometimes causes Github's diff viewer to 500.

I'll continue running this against my cluster and see if I can find any other edge cases.

luna-duclos commented:
How has this been behaving on your deployment ?

euank commented:
Working well as best I can tell.

I also double checked that it behaved well for a short api-server downtime (e.g. monitorCertificates/Ingresses and getCertificates/etc still come back).

I haven't really exercised gc at all, but I've had it renew and create certs just fine.

luna-duclos commented on processor.go:
```diff
@@ -364,8 +361,8 @@ func (p *CertProcessor) processCertificate(cert Certificate) (processed bool, er
-		// If certificate expires in more than a week, don't renew
-		if parsedCert.NotAfter.After(time.Now().Add(time.Hour * 24 * 7)) {
+		// If certificate expires in more than 20 days, don't renew
+		if parsedCert.NotAfter.After(time.Now().Add(time.Hour * 24 * 20)) {
```
Seems out of place in this PR, was this accidently added perhaps ?

euank replied:
Good catch.

luna-duclos reviewed (approved):
Just one small comment here, looks good.

euank commented:
One thing to note is that this will conflict with #46, so whoever's second will have to resolve some conflicts. Since that one came in first, it might make sense to merge that first and then I'll rebase again.

luna-duclos commented:
This can be rebased.

euank committed:
vendor: update k8s/client-go

euank committed:
k8s: move to client-go package

This makes a minimal set of changes to migrate to the client-go package
over. Most importantly, this gets rid of the abomination that was
'apihost' and replaces it with InClusterConfig.

This is not intended to change any functionality. This includes not
changing in any significant way the endpoints that are hit, when, and
why.

euank committed:
kubectl-proxy: remove in favor of client-go

euank committed:
processor: delete dead code

euank committed:
k8s: update monitoring to use cache pkg

This code is a bit cleaner and hopefully more robust.

euank committed:
k8s: use a token bucket rather than fixed 5s sleep

euank commented:
Rebased
